### PR TITLE
Use alternate screen and refactor terminal output.

### PR DESCRIPTION
**Pull request by frozolotl**
Prior to this PR, the entire terminal was cleared when running `typst watch`.
This is somewhat disruptive to the user experience.
This PR changes that behavior and only clears what Typst has written.

Because to my knowledge there isn't an easy to use ANSI escape sequence for this, I centralized all terminal output in one struct: `TermOut`.
It records the number of lines written and can use that to clear that amount if requested.
This also has the nice benefit of removing certain escape sequence machinery from unrelated functions.

I don't particularly like having to pass this `TermOut` struct everywhere, but it's the best solution I could come up with.
Neither do I like it relying on a `Rc<RefCell<_>>` type thing, but it appears to be necessary due to the way `typst::World` works.

Overall though I am happy with it and I think it is totally usable.

### CODE REVIEW

**Comment by mattfbacon** [comment]
I think the "proper" solution to this is to use the alternate screen.

**Comment by alerque** [comment]
I actually like the screen clear. There is precedent for this with other similar terminal interfaces (e.g. `watch`). And because the length of output from this can vary (as when Typst throws errors) this keeps a consistent starting point on the screen even as recompilations might change the length of the output.

**Comment by frozolotl** [comment]
Copy-paste from the latest commit:

Due to popular demand, I changed the behavior as follows:
By default, fullscreen mode is used. It differs from the implementation on main in that it uses an alternate screen.
That means, once you exit `typst watch`, the prior terminal state reappears.
If enabled via --no-fullscreen, the terminal is not cleared and no alternative screen appears.

What I don't like about this implementation:
- It is more complex because it requires signal handling.
- In fullscreen mode, the last compilation output disappears on exit.
  I might change that in the near future.

Feedback on this would be nice, as this appears to be a more opinionated issue than I expected.

**Comment by mattfbacon** [comment]
This is interesting because what you described about the compilation output disappearing is exactly what I would expect. And indeed `watch` seems to implement this exact behavior.

As for the signal handling, that's unfortunate but I think it's unavoidable.

**Comment by frozolotl** [comment]
> This is interesting because what you described about the compilation output disappearing is exactly what I would expect. And indeed `watch` seems to implement this exact behavior.
After thinking a bit more about it, yes, you are right. It _would_ be a bit strange if after exiting it simply reappeared again, just moved a bit down to whereever the command was executed in the shell.

**Comment by laurmaedje** [comment]
The alternate screen makes total sense, but I don't really understand the use case for `--no-fullscreen` yet. Could you expand a bit on this?

**Comment by frozolotl** [comment]
> The alternate screen makes total sense, but I don't really understand the use case for `--no-fullscreen` yet. Could you expand a bit on this?

It would enable users to view the shell history without having to exit `typst watch`. I can't give a specific example where this would be useful though.
For me it mainly seemed like an easy-to-implement flag that users might want to enable because they prefer that behavior. Probably by defining a shell alias.
I like it a bit more than clearing the terminal, but I'm not very attached to it either.

By removing it, the `WatchCommand` struct could be removed again and we wouldn't need to track the number of lines written. The rest would stay the same.

**Comment by mattfbacon** [comment]
I think users would also be able to go back to the main screen temporarily with Ctrl-Z (sigstop).

**Comment by laurmaedje** [comment]
I'm leaning towards just supporting the alternate screen and not the special behaviour that requires line tracking (and thus the `TermOut` construct).

**Comment by laurmaedje** [comment]
@frozolotl do you still plan on continuing work on this?

**Comment by laurmaedje** [comment]
I'd like to see the `TermOut` construct removed before merging. No rush though from my side.

**Comment by frozolotl** [comment]
I have finally done some cleanup and rebased the whole thing to be up-to-date so that it can be merged without any merge conflicts.

I decided to keep `TermOut` because it is actually pretty useful as we still need to have some state for the alternate screen.
However, now there is no longer a shared instance that has to be passed around, but instead the function `terminal::out()` can be called to get a handle to it.

**Comment by mattfbacon** [comment]
This is a nit but why do we need an Arc if it's in a `static`? Can't we use `&'static TermOutInner` instead?

**Comment by frozolotl** [comment]
> This is a nit but why do we need an Arc if it's in a `static`? Can't we use `&'static TermOutInner` instead?

Good point. Yesterday I only thought of removing the separation of `TermOut` and `TermOutInner`, for which I didn't find a solution.
But simply keeping `TermOut` and using a `&'static TermOutInner` does work.
Thanks!

File: crates/typst-cli/src/compile.rs [hunk 1]
```diff
@@ -3,11 +3,12 @@ use std::path::{Path, PathBuf};
 
 use chrono::{Datelike, Timelike};
 use codespan_reporting::diagnostic::{Diagnostic, Label};
-use codespan_reporting::term::{self, termcolor};
+use codespan_reporting::term;
 use ecow::{eco_format, EcoString};
 use parking_lot::RwLock;
-use rayon::iter::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIterator};
-use termcolor::{ColorChoice, StandardStream};
+use rayon::prelude::{
```
Comment (laurmaedje) [inline crates/typst-cli/src/compile.rs:9]: Why did the rayon import change?

File: crates/typst-cli/src/compile.rs [hunk 2]
```diff
@@ -308,16 +309,11 @@ fn open_file(open: Option<&str>, path: &Path) -> StrResult<()> {
 
 /// Print diagnostic messages to the terminal.
 pub fn print_diagnostics(
-    world: &SystemWorld,
+    world: &mut SystemWorld,
```
Comment (laurmaedje) [inline crates/typst-cli/src/compile.rs:312]: why this change?

File: crates/typst-cli/src/terminal.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+/// A utility that allows users to write colored terminal output.
+/// If colors are not supported by the terminal, they are disabled.
+/// This type also allows for deletion of previously written lines.
+#[derive(Clone)]
+pub struct TermOut {
+    inner: &'static TermOutInner,
+}
+
+impl TermOut {
```
Comment (laurmaedje) [inline crates/typst-cli/src/terminal.rs:52]: why do some of `TermOut`'s methods take `&mut self`?

also, why do we have `TermOut` and `TermOutInner` instead of just letting `out()` return an `&'static TermOut`?

edit: probably because Write requires `&mut self`? perhaps `out()` could return a locked TermOut like `std::io::stdout()` does?

File: crates/typst-cli/src/main.rs [hunk 4]
```diff
@@ -33,14 +35,21 @@ static ARGS: Lazy<CliArguments> = Lazy::new(CliArguments::parse);
...
         Command::Compile(command) => crate::compile::compile(timer, command.clone()),
         Command::Watch(command) => crate::watch::watch(timer, command.clone()),
         Command::Query(command) => crate::query::query(command),
         Command::Fonts(command) => crate::fonts::fonts(command),
         Command::Update(command) => crate::update::update(command),
     };
 
+    // Leave the alternate screen if it was opened. This operation is done here
+    // so that it is executed prior to printing the final error.
+    res = res.or(terminal::out()
```
Comment (laurmaedje) [inline crates/typst-cli/src/main.rs:49]: This or confused me for a bit. Perhaps, this should just be `let res2 = terminal::out()...` and then `if let Err(msg) = res.or(res2)`.

Or maybe we can use expect here. If we can't leave the alternative screen, will print_error not also most probably just fail and then we get "failed to print error" rather than the more helpful "failed to leave alternate screen" panic?

File: crates/typst-cli/src/terminal.rs (hunk 3 above)
Comment (mattfbacon) [reply crates/typst-cli/src/terminal.rs:52]: I agree with the suggestion to return a locked handle. Locking the mutex in every write call is sub-optimal.

File: crates/typst-cli/src/main.rs (hunk 4 above)
Comment (mattfbacon) [reply crates/typst-cli/src/main.rs:49]: Also, should be `or_else` for laziness if we keep it.

Comment (laurmaedje) [reply crates/typst-cli/src/main.rs:49]: That was also my first thought, but I think it is intentionally not or_else because we want to leave the alternative screen even if `res` contains an error.

Comment (mattfbacon) [reply crates/typst-cli/src/main.rs:49]: I see, it's definitely confusing then.

File: crates/typst-cli/src/compile.rs (hunk 1 above)
Comment (frozolotl) [reply crates/typst-cli/src/compile.rs]: This appears to be a mistake that occured while rebasing.

File: crates/typst-cli/src/compile.rs (hunk 2 above)
Comment (frozolotl) [reply crates/typst-cli/src/compile.rs]: This appears to have been a mistake that occured while rebasing.

File: crates/typst-cli/src/terminal.rs (hunk 3 above)
Comment (frozolotl) [reply crates/typst-cli/src/terminal.rs:52]: > why do some of `TermOut`'s methods take `&mut self`?

I thought it would be clearer which methods mutate things like that. It doesn't change anything in actual usage.

> also, why do we have `TermOut` and `TermOutInner` instead of just letting `out()` return an `&'static TermOut`?
> 
> edit: probably because Write requires `&mut self`? perhaps `out()` could return a locked TermOut like `std::io::stdout()` does?

Yes, it's necessary because of `Write's` and `WriteColor`s `&mut self` parameter.
`std::io::stdout()` does not actually return a locked handle. One needs to call `.lock()` on it.

I'm not sure always locking is benefitial here:
- The write methods are only rarely called, so the performance benefit would be neglegible.
- The `.is_active()` method (among others) does not require locking. It is very regularly called. It's not a huge performance hit to lock here, but not very nice either.
- Locking on every `terminal::out()` call might cause a deadlock if insufficient care is taken when writing the code.

The alternative would be an extra `.lock()` method like `std::io::stdout()` and `termcolor::StandardStream` offer, but I don't think there is enough of a case to warrant the extra code required.

I therefore think it is preferable to implicitly lock in every write call, even though it does add a small performance overhead.

Comment (laurmaedje) [reply crates/typst-cli/src/terminal.rs:52]: Okay

**Comment by laurmaedje** [comment]
I've only just properly noticed the exit handler logic. Does that mean that an ongoing slow `typst watch` compile cannot be terminated with Ctrl+C (i.e. it only terminates once the compilation finishes)?

**Comment by frozolotl** [comment]
> I've only just properly noticed the exit handler logic. Does that mean that an ongoing slow `typst watch` compile cannot be terminated with Ctrl+C (i.e. it only terminates once the compilation finishes)?

Good point. I have now changed it so that first, `active` is set to false and the watch loop can exit normally. This would ensure that destructors are run and files written.
However, if that takes too long (750 ms at the moment), then the process is forcibly terminated.

**Comment by laurmaedje** [comment]
Thank you!
